src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisaulStudioFileChangeTracker.cs

```diff
@@ -58,71 +71,109 @@ public override void StartListening()
         {
             _foregroundDispatcher.AssertForegroundThread();
 
-            try
+            if (_fileChangeAdviseTask != null)
             {
-                if (_fileChangeCookie == VSConstants.VSCOOKIE_NIL)
-                {
-                    var hr = _fileChangeService.AdviseFileChange(
-                        FilePath,
-                        FileChangeFlags,
-                        this,
-                        out _fileChangeCookie);
-
-                    Marshal.ThrowExceptionForHR(hr);
-                }
+                // Already listening
+                return;
             }
-            catch (Exception exception)
+
+            if (_fileChangeUnadviseTask?.IsCompleted == false)
             {
-                _errorReporter.ReportError(exception);
+                // An unadvise operation is still processing, block the foreground thread until it completes.
+                _fileChangeUnadviseTask.Join();
             }
+
+            _fileChangeAdviseTask = _joinableTaskFactory.RunAsync(async () =>
+            {
+                if (_fileChangeUnadviseTask != null)
```

Comment:
Based on line 83, wouldn't the unadvise task always have completed by the time you get here? Is this when a Stop is invoked while start's still processing? If so, this needs a comment explaining the behavior.

Reply:
Ahh, oops must not have committed the removal of this. This deadlocks in race scenarios.